clippy_lints/src/try_err.rs

```diff
@@ -0,0 +1,120 @@
+    ///     if fail {
+    ///       return Err("failed".into());
+    ///     }
+    ///     Ok(0)
+    /// }
+    /// ```
+    pub TRY_ERR,
+    style,
+    "return errors explicitly rather than hiding them behind a `?`"
+}
+
+declare_lint_pass!(TryErr => [TRY_ERR]);
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for TryErr {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        // Looks for a structure like this:
+        // match ::std::ops::Try::into_result(Err(5)) {
+        //     ::std::result::Result::Err(err) =>
+        //         #[allow(unreachable_code)]
+        //         return ::std::ops::Try::from_error(::std::convert::From::from(err)),
+        //     ::std::result::Result::Ok(val) =>
+        //         #[allow(unreachable_code)]
+        //         val,
+        // };
+        if_chain! {
+            if let ExprKind::Match(ref match_arg, _, MatchSource::TryDesugar) = expr.node;
+            if let ExprKind::Call(ref match_fun, ref try_args) = match_arg.node;
+            if let ExprKind::Path(ref match_fun_path) = match_fun.node;
+            if match_qpath(match_fun_path, &["std", "ops", "Try", "into_result"]);
+            if let Some(ref try_arg) = try_args.get(0);
+            if let ExprKind::Call(ref err_fun, ref err_args) = try_arg.node;
+            if let Some(ref err_arg) = err_args.get(0);
+            if let ExprKind::Path(ref err_fun_path) = err_fun.node;
+            if match_qpath(err_fun_path, &paths::RESULT_ERR);
+            if let Some(return_type) = find_err_return_type(cx, &expr.node);
+
+            then {
+                let err_type = cx.tables.expr_ty(err_arg);
+                let suggestion = if err_type == return_type {
+                    format!("return Err({})", snippet(cx, err_arg.span, "_"))
+                } else {
+                    format!("return Err({}.into())", snippet(cx, err_arg.span, "_"))
+                };
+
+                span_lint_and_then(
+                    cx,
+                    TRY_ERR,
+                    expr.span,
+                    &format!("confusing error return, consider using `{}`", suggestion),
+                    |db| {
+                        db.span_suggestion(
+                            expr.span,
+                            "try this",
+                            suggestion,
+                            Applicability::MaybeIncorrect
+                        );
+                    },
+                );
+            }
+        }
+    }
+}
+
+// In order to determine whether to suggest `.into()` or not, we need to find the error type the
+// function returns. To do that, we look for the From::from call (see tree above), and capture
+// its output type.
+fn find_err_return_type<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, expr: &'tcx ExprKind) -> Option<Ty<'tcx>> {
+    if let ExprKind::Match(_, ref arms, MatchSource::TryDesugar) = expr {
+        arms.iter().filter_map(|ty| find_err_return_type_arm(cx, ty)).nth(0)
```

Comment:
Couldn't this be written as
```suggestion
        arms.iter().find_map(|ty| find_err_return_type_arm(cx, ty))
```

If so, this should be added as a lint to `methods/mod.rs`